File: Practica3/Marco teorico/4. LSP/LSP.cpp
#include <iostream>
using namespace std;
class Animal{
public:
 virtual void comer(){
   cout << "El animal está comiendo." << endl;
 }
};




class Mamifero : public Animal{
public:
 virtual void amamantar(){
   cout << "El mamífero está amamantando." << endl;
 }
};




class Ave : public Animal{
public:
 virtual void volar(){
   cout << "El ave está volando." << endl;
 }
};




class Perro : public Mamifero {
public:
 virtual void ladrar() {
 cout << "El perro está ladrando." << endl;
 }
};




class Pajaro : public Ave{
public:
 virtual void cantar(){
   cout << "El pájaro está cantando." << endl;
 }
};

int main()
{
   Animal* animal1 = new Animal();
   Mamifero* mamifero1 = new Mamifero();
   Ave* ave1 = new Ave();
   Perro* perro1 = new Perro();
   Pajaro* pajaro1 = new Pajaro();
   animal1->comer();
   mamifero1->comer();
   mamifero1->amamantar();
   ave1->comer();
   ave1->volar();
   perro1->comer();
   perro1->amamantar(); // Esto no tiene sentido en un perro
   perro1->ladrar();
   pajaro1->comer();
   pajaro1->volar();
   pajaro1->cantar();
   delete animal1;
   delete mamifero1;
   delete ave1;
   delete perro1;
   delete pajaro1;
return 0;
}

File: Practica3/Marco teorico/4. LSP/LSP.py
class Animal:
    def comer(self):
        print("El animal está comiendo.")


class Mamifero(Animal):
    def amamantar(self):
        print("El mamífero está amamantando.")


class Ave(Animal):
    def volar(self):
        print("El ave está volando.")


class Perro(Mamifero):
    def ladrar(self):
        print("El perro está ladrando.")


class Pajaro(Ave):
    def cantar(self):
        print("El pájaro está cantando.")



animal1 = Animal()
mamifero1 = Mamifero()
ave1 = Ave()
perro1 = Perro()
pajaro1 = Pajaro()

animal1.comer()
mamifero1.comer()
mamifero1.amamantar()
ave1.comer()
ave1.volar()
perro1.comer()
perro1.ladrar()
pajaro1.comer()
pajaro1.volar()
pajaro1.cantar()